plugins/thunderbolt/fu-thunderbolt-image.c

```diff
@@ -235,6 +247,56 @@ read_ucode_section_len (guint32                       offset,
 	return TRUE;
 }
 
+/*
+ * reads generic entries from DROM based on type field and fills
+ * location to point to the entry data if found
+ */
+static gboolean
+read_drom_entry_location (const FuThunderboltFwObject *fw,
+			  guint8                       type,
+			  FuThunderboltFwLocation     *location,
+			  GError                     **error)
+{
+	const FuThunderboltFwLocation drom_len_loc = { .offset = 0x0E, .len = 2, .section = DROM_SECTION, .description = "DROM length" };
+	FuThunderboltFwLocation drom_entry_loc = { .len = 2, .section = DROM_SECTION, .description = "DROM generic entry" };
+	guint16 drom_size;
+
+	if (!read_uint16 (&drom_len_loc, fw, &drom_size, error))
+		return FALSE;
+
+	drom_size &= 0x0FFF;
+	/* drom_size is size of DROM block except for identification
+	 * section and crc32 so add them here */
+	drom_size += 9 + 4;
+
+	/* DROM entries start right after the identification section */
+	drom_entry_loc.offset = 9 + 4 + 9;
+
+	do {
+		g_autoptr(GByteArray) entry;
+		guint8 entry_type;
+		guint8 entry_length;
+
+		entry = read_location (&drom_entry_loc, fw, error);
+		if (entry == NULL)
+			return FALSE;
+
+		entry_length = entry->data[0];
+		entry_type = entry->data[1] & 0x3F;
+
+		/* generic entry (port bit is not set) */
+		if ((entry->data[1] & (1 << 7)) == 0 && entry_type == type) {
+			location->len = entry_length - 2;
+			location->offset = drom_entry_loc.offset + 2;
+			return TRUE;
+		}
+
+		drom_entry_loc.offset += entry_length;
+	} while (drom_entry_loc.offset < drom_size);
+
+	return FALSE;
```

Comment:
I think you'll need to explicitly set the error in this `FALSE` scenario.

Reply:
Hmm, it is not an error not to have DROM generic entry found. This function is used to fill in an optional location (Multi Controller) entry but it is fine if it does not exists in the controller NVM. The error is only set if we failed to read the DROM for some reason.